internal/handlers/common/projection.go

```diff
@@ -221,28 +221,270 @@ func projectDocumentWithoutID(doc *types.Document, projection *types.Document, i
+// with the same key. Dot notation with array index path does not include
+// the field unlike document.SetByPath(path).
+// Inclusion projection with non-existent path creates an empty document
+// or an empty array based on what source has.
+//
+//	Example: "v.foo" path inclusion projection:
+//	{v: {foo: 1, bar: 1}}               -> {v: {foo: 1}}
+//	{v: {bar: 1}}                       -> {v: {}}
+//	{v: [{bar: 1}]}                     -> {v: [{}]}
+//	{v: [{foo: 1}, {foo: 2}, {bar: 1}]} -> {v: [{foo: 1}, {foo: 2}, {}]}
+//
+//	Example: "v.0.foo" path inclusion projection:
+//	{v: [{foo: 1}, {foo: 2}, {bar: 1}]} -> {v: [{}, {}, {}]}
+func includeProjection(path types.Path, source any, projected *types.Document) (*types.Array, error) {
+	key := path.Prefix()
+
+	switch source := source.(type) {
+	case *types.Document:
+		embeddedSource, err := source.Get(key)
+		if err != nil {
+			// key does not exist, nothing to set.
+			return nil, nil
+		}
+
+		if path.Len() <= 1 {
+			// path reached suffix, set field in projected.
+			setBySourceOrder(key, embeddedSource, source, projected)
+			return nil, nil
+		}
+
+		doc := new(types.Document)
+
+		if projected.Has(key) {
+			// set doc if projected has field from other projection field.
+			v := must.NotFail(projected.Get(key))
+			if d, ok := v.(*types.Document); ok {
+				doc = d
+			}
+		}
+
+		// when next prefix has an arr use arr,
+		// if it has a document, doc is set by includeProjection.
+		arr, err := includeProjection(path.TrimPrefix(), embeddedSource, doc)
+
+		switch {
+		case errors.Is(err, pathNotExists):
+			// path does not exist, nothing to set.
+			return nil, nil
+		case err != nil:
+			return nil, err
+		}
+
+		if isEmpty(key, projected) {
+			// only set projected if projected contains empty value.
+			// If projected is `{v: {foo: 1}}` and arr is `{}`,
+			// do not overwrite it.
+			if _, ok := embeddedSource.(*types.Array); ok {
+				// set arr to projected
+				projected.Set(key, arr)
+				return nil, nil
+			}
+
+			setBySourceOrder(key, doc, source, projected)
+		}
+
+		return nil, nil
+	case *types.Array:
+		iter := source.Iterator()
+		defer iter.Close()
+
+		arr := new(types.Array)
+
+		for {
+			_, arrElem, err := iter.Next()
+			if err != nil {
+				if errors.Is(err, iterator.ErrIteratorDone) {
+					break
+				}
+
+				return nil, lazyerrors.Error(err)
+			}
+
+			if _, ok := arrElem.(*types.Document); !ok {
+				continue
+			}
+
+			doc := new(types.Document)
+			if _, err = includeProjection(path, arrElem, doc); err != nil {
+				return nil, err
+			}
+
+			arr.Append(doc)
+		}
+
+		return arr, nil
+	default:
+		return nil, pathNotExists
+	}
+}
+
+// isEmpty returns true if projected document at key is empty. It's empty if:
+// - key does not exist in projected document,
+// - projected contains empty document at key, or
+// - projected contains an array which contain empty documents at key.
```

Comment:
To make it render as list
```suggestion
//   - key does not exist in projected document,
//   - projected contains empty document at key, or
//   - projected contains an array which contain empty documents at key.
```